### PR TITLE
Fix wp_get_raw_referer() to make sure it always returns strings/booleans

### PR DESCRIPTION
**Pull request by xknown**
Trac ticket: https://core.trac.wordpress.org/ticket/57670#ticket

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Review by johnbillion — APPROVED** [review]

**Review by jrfnl — COMMENTED** [review]
The patch looks good to me.

I am left wondering about two things though:
1. Is the `_wp_http_referer` REQUEST variable documented anywhere ? And if so, is that documentation clear enough that this should always be set to a `string` ? And if not documented, is there anywhere we can add documentation about this ?
2. Can we think of any legitimate reasons for the variable to be set to an array ?
    The only thing I can come up with is a redirect-chain.

**Comment by xknown** [comment]
> The patch looks good to me.
> 
> I am left wondering about two things though:
> 
>     1. Is the `_wp_http_referer` REQUEST variable documented anywhere ? And if so, is that documentation clear enough that this should always be set to a `string` ? And if not documented, is there anywhere we can add documentation about this ?
> 
>     2. Can we think of any legitimate reasons for the variable to be set to an array ?
>        The only thing I can come up with is a redirect-chain.

I don't think it's a documentation problem, or developers using it on the wrong way. We noticed this issue because there were lots of PHP fatals related to people running automated security tools.

**Comment by jrfnl** [comment]
> > 2. Can we think of any legitimate reasons for the variable to be set to an array ?
> >    The only thing I can come up with is a redirect-chain.
> 
> I don't think it's a documentation problem, or developers using it on the wrong way. We noticed this issue because there were lots of PHP fatals related to people running automated security tools.

Well, servers don't set that variable by default, so it gets set somewhere.... whether it is by WP Core, by a plugin or by an automated security tool. So, what I'm really asking is, can we identify in some way, what code is setting the variable to an array and for what purpose.
Based on _this_ function's documentation, setting the variable to an array is _doing it wrong_, but I'm wondering if there are legitimate reasons for that and what the source is (and if it is Core, we need to fix Core or decide that arrays are allowed).

File: tests/phpunit/tests/functions/referer.php [hunk 1]
```diff
@@ -156,4 +156,12 @@ public function test_raw_referer_both() {
 		$_REQUEST['_wp_http_referer'] = addslashes( 'http://foo.bar/baz' );
 		$this->assertSame( 'http://foo.bar/baz', wp_get_raw_referer() );
 	}
+
+	/**
+	 * @ticket 57670
+	 */
+	public function test_raw_referer_is_empty_on_invalid_request_parameter() {
```
Comment (costdev) [inline tests/phpunit/tests/functions/referer.php]:
```suggestion
	public function test_raw_referer_is_false_on_invalid_request_parameter() {
```

**Comment by azaozz** [comment]
Yep, more/better docs are always a big plus! The question is: where would be a good place to add them :)

> servers don't set that variable by default, so it gets set somewhere...

Right. It gets set in few places, mostly through `wp_nonce_field()` (which is used a lot) that calls `wp_referer_field()` that returns the markup for a hidden HTML input element. 

Seems the docblock for `wp_referer_field()` is the most appropriate place although not sure how "visible" it would be as that function is rarely used on its own. So perhaps can add the docs in two places: the docblocks for `wp_referer_field()` and for `wp_get_raw_referer()`?

**Comment by johnbillion** [comment]
https://core.trac.wordpress.org/changeset/56115

**Pull request closed without merging**